context/app/api/client.py

```diff
@@ -152,6 +145,40 @@ def get_vitessce_conf_cells_and_lifted_uuid(self, entity):
             return VitessceConfLiftedUUID(ConfCells({'error': message}, None), None)
 
 
+def _make_query(constraints):
+    '''
+    Given a constraints dict of lists,
+    return a ES query that handles all structual variations.
+    Repeated values for a single key are OR;
+    Separate keys are AND.
+
+    >>> constraints = {'color': ['red', 'green'], 'number': ['42']}
+    >>> query = _make_query(constraints)
+    >>> from pprint import pp
+    >>> pp(query['bool'])
+    {'must': [{'bool': {'should': [{'term': {'metadata.metadata.color.keyword': 'red'}},
+                                   {'term': {'mapped_metadata.color.keyword': 'red'}},
+                                   {'term': {'metadata.metadata.color.keyword': 'green'}},
+                                   {'term': {'mapped_metadata.color.keyword': 'green'}}]}},
+              {'bool': {'should': [{'term': {'metadata.metadata.number.keyword': '42'}},
+                                   {'term': {'mapped_metadata.number.keyword': '42'}}]}}]}
+    '''
+    shoulds = [
+        [
+            {"term": {f'{root}.{k}.keyword': v}}
+            for v in v_list
+            for root in ['metadata.metadata', 'mapped_metadata']
```

Comment:
Do we expect the values for the properties in `metadata.metadata` and `mapped_metadata` to be the same?

Reply:
Or is that the purpose of the `should`?

Reply:
yeah: The JSON document structure is a mess, and the "metadata" is in different locations for different entity types. The `should` (ie, `OR`) handles this: check all the possible paths, and see if it's under any of them.